pkg/applicationserver/applicationserver.go

```diff
@@ -780,255 +728,65 @@ func (as *ApplicationServer) handleJoinAccept(ctx context.Context, ids ttnpb.End
 				StartedAt: time.Now().UTC(),
 			}
 			mask = append(mask, "pending_session")
-			if !as.skipPayloadCrypto(ctx, link, dev, dev.PendingSession) {
-				if len(joinAccept.InvalidatedDownlinks) > 0 {
-					// The Network Server does not reset the downlink queues as the new security session is established,
-					// but rather when the session is confirmed on the first uplink. The downlink queue of the current
-					// session is passed as part of the join-accept in order to allow the Application Server to compute
-					// the downlink queue of this new pending session.
-					logger := logger.WithField("count", len(joinAccept.InvalidatedDownlinks))
-					logger.Debug("Recalculating downlink queue to restore downlink queue on join")
-					if err := as.recalculatePendingDownlinkQueue(ctx, dev, link, previousSession, joinAccept.InvalidatedDownlinks); err != nil {
-						logger.WithError(err).Warn("Failed to recalculate downlink queue; items lost")
-					}
-					joinAccept.InvalidatedDownlinks = nil
+			if as.skipPayloadCrypto(ctx, link, dev, dev.PendingSession) {
+				return dev, mask, nil
+			}
+			joinAccept.AppSKey = nil
+			if len(joinAccept.InvalidatedDownlinks) == 0 {
+				return dev, mask, nil
+			}
+
+			// The Network Server does not reset the downlink queues as the new security session is established,
+			// but rather when the session is confirmed on the first uplink. The downlink queue of the current
+			// session is passed as part of the join-accept in order to allow the Application Server to compute
+			// the downlink queue of this new pending session.
+			logger := logger.WithField("count", len(joinAccept.InvalidatedDownlinks))
+			logger.Debug("Recalculating downlink queue to restore downlink queue on join")
+
+			items := make([]*ttnpb.ApplicationDownlink, 0, len(joinAccept.InvalidatedDownlinks))
+			for _, msg := range joinAccept.InvalidatedDownlinks {
+				if err := as.decryptDownlink(ctx, dev, msg, previousSession); err != nil {
+					logger.WithError(err).Warn("Failed to decrypt downlink message; drop item")
+					registerDropDownlink(ctx, ids, msg, err)
+					continue
 				}
-				joinAccept.AppSKey = nil
+				items = append(items, msg)
+			}
+			joinAccept.InvalidatedDownlinks = nil
+			if len(items) == 0 {
+				return dev, mask, nil
 			}
+
+			if session := dev.Session; session != nil {
+				session.LastAFCntDown = items[0].FCnt - 1
```

Comment:
Why is this? I thought that `LastAFCntDown` would be the last used `FCntDown`, i.e. `items[len(items)-1].FCnt`.

BTW, are we sure `FCnt` is never `0`?

Reply:
I had a long comment here, which I will leave out for historical reasons, but the TLDR is probably that it would be better to just push the uplinks only to the pending session, if possible. I'll check out if I can do it cleanly.

---

There are two mechanisms here, which actually probably should end up as godoc.

1. We call them invalidated downlinks, but the queue itself has not been cleaned for the `dev.Session` -  we just receive a copy. See https://github.com/TheThingsNetwork/lorawan-stack/blob/c0e437afc7fccb9d506e05f73b4a2d2c3db54d13/pkg/networkserver/downlink.go#L1667-L1690

2. `as.AttempDownlinkQueueOp` will replace both queues (as you cannot replace only one queue). As such, it has to compute the FCnt of each uplink again. If we don't lower this `FCnt` before calling it, each time a device joins, the `FCnt` of each element will increase by the queue size. 

Take this example:
- You schedule 3 downlinks to your device, with FCnts 42, 43, 44. `LastAFCntDown` is 44.
- Your device rejoins. We receive these downlinks in `InvalidatedDownlinks`. 
- We have to do a `replace` operation in order to create the queue for the pending session.
- When we calculate the downlinks for `dev.Session`, we start from 44, so we'll have 45, 46, 47.

If your device rejoins too many times at some point we will run out of FCnts. As such, we attempt to keep the `LastAFCntDown` steady.

Reply:
> * You schedule 3 downlinks to your device, with FCnts 42, 43, 44. `LastAFCntDown` is 44.
> [...]
> * When we calculate the downlinks for `dev.Session`, we start from 44, so we'll have 45, 46, 47.
> 
> If your device rejoins too many times at some point we will run out of FCnts. As such, we attempt to keep the `LastAFCntDown` steady.

Wait, the `LastAFCntDown` survives a session? That doesn't work with rollovers in the frame.

Reply:
No - the new `dev.PendingSession` starts with `LastAFCntDown=0`, and then we push the old downlink queue to this pending session. So if you had `n` downlinks in `dev.Session`, your `dev.PendingSession` will contain `n` downlinks, with framecounters between 1 and n, and `LastAFCntDown=n`.

The behavior that I described where the counter keeps increasing for `dev.Session` is no longer applicable in any shape or form, after 33c6276310b003a1a18f9c63706e27978451e976. Now on `JoinAccept` we don't even touch the downlink queue of `dev.Session`, we just create a new one for `dev.PendingSession`, for which the counter starts at 0.